README.rst
=====
OView
=====

OView is a multi-purpose image and data viewer for Omega hdf files.

The vision is to implement a core of reading and viewing functionality, and to
be extensible for supporting new Omega-60 or Omega-EP hdf data sources. Some
rudimentary data processing and analysis will be builtin, more in-depth
analysis projects may prefer to extend with subclassing or import.

**Supported file types**
------------------------

builtin support for the following file types:

- image files (.jpg, .png, .bmp, .tif)
- a subset of .hdf files from the OMEGA database:
  - p510
  - asbo
  - pxrdip

** Dependencies **
------------------

oview was last developed with the following environment:

- python 3.6 (earlier versions unverified)
- numpy
- scipy
- PyQt5
- Qt5
- qtpy
- pyqtgraph

Additional (optional) python packages for full functionality:

- matplotlib
- pyhdf



HISTORY.rst
.. :changelog:

oview version history
---------------------

0.2.3 (2017-05-25)
++++++++++++++++++

+ add preliminary sop support

0.2.2 (2017-05-24)
++++++++++++++++++

+ fixed crash when no file selected in openfile dialog
+ rescale aspect ratio when opening file


0.2.1 (2017-05-22)
++++++++++++++++++

+ restructuring for intended long-term development
+ initial support for Omega p510 summary .hdf files
+ initial github commit


0.2.0 (2017-04-17)
++++++++++++++++++

+ support for some Omega/Omega-EP .hdf files: PXRDIP, ASBO


0.1.0 (2016-12-13)
++++++++++++++++++

+ basic support for .jpg, .png, .tif, .bmp files


oview/oview.py
"""OView: Multi-purpose image and data viewer for Omega data files.

keyboard shortcuts:
    F1: help window
    l: open look-up-table window
    o: open image file (supported formats: hdf, tif, png, jpg, bmp)
    h: reset plot limits
    Esc: close window
""" # module docstring doubles as help window text

__author__ = "J. Ryan Rygg"
__version__ = "0.2.2"

#==============================================================================
import os.path as osp
import sys

import numpy as np
import pyqtgraph as pg
from pyqtgraph.Qt import QtCore, QtGui
from PyQt5.QtWidgets import (
        QApplication, QMainWindow, QDesktopWidget, QFileDialog,
        QLabel, QTextEdit,
)
from PyQt5.QtGui import QIcon

from .io import read_hdf

# lazy import: won't import until needed (gives tiny speedup on startup)
sp_misc = None # import scipy.misc

#==============================================================================
# Interpret image data as row-major instead of col-major
pg.setConfigOptions(imageAxisOrder='row-major')

FILE_PATH = osp.abspath(osp.dirname(__file__))
ICON_PATH = osp.abspath(osp.join(FILE_PATH, '..', 'oview_data', 'oview48.png'))
STARTDIR = None # TODO: allow user to specify default starting directory

#==============================================================================
class OViewMainWindow(QMainWindow):
    """OView MainWindow"""
    WINTITLE_PREFIX = "OView"

    def __init__(self, filename=None):
        QMainWindow.__init__(self)
        self.setAttribute(QtCore.Qt.WA_DeleteOnClose)
        self._init_UI()
        self._init_shortcuts()
        self.show()

        if filename is None:
            self.openDialog()
        else:
            self.open(filename)

        self.lut.show()
        self.lut.move_to_parent()
        self.activateWindow() # return active window state to self

    def _init_UI(self):
        self.setWindowTitle(self.WINTITLE_PREFIX)
        self.setWindowIcon(QIcon(ICON_PATH))
        self.resize_maxrect()

        #--- pyqtgraph widgets ---
        self.graphics_widget = pg.GraphicsLayoutWidget()
        self.graphics_widget.ci.layout.setContentsMargins(2, 2, 2, 2)
        self.setCentralWidget(self.graphics_widget)

        self.statusbar_right = QLabel("")
        self.statusBar().addPermanentWidget(self.statusbar_right)

        # A plot area (ViewBox + axes) for displaying the image
        self.plot1 = self.graphics_widget.addPlot()
        self.plot1.setAspectLocked(True)

        # Item for displaying image data
        self.img1 = pg.ImageItem()
#        self.img1.setImage(np.ones((2,2)))
        self.plot1.addItem(self.img1)

        # Contrast/color control
        self.lut = LUTWindow(self)
        self.hist = self.lut.hist

        # mouse move updates
        self.proxy = pg.SignalProxy(self.plot1.scene().sigMouseMoved,
                                    rateLimit=60, slot=self.mouse_moved)

    def _init_shortcuts(self):
        """initialize keyboard shortcuts"""
        QShortcut = QtGui.QShortcut
        Qt = QtCore.Qt

        QShortcut(Qt.Key_F1, self, self.helpDialog)
        QShortcut(Qt.Key_O, self, self.openDialog)
        QShortcut(Qt.Key_L, self, self.lutDialog)
        QShortcut(Qt.Key_1, self, self.reset_plotview)
        QShortcut(Qt.Key_H, self, self.reset_plotview)

        # add various shortcuts to close the window
        QShortcut(Qt.Key_Escape, self, self.close)
        for ks in ("Ctrl+w", "Ctrl+F4", "Ctrl+q"):
            QShortcut(QtGui.QKeySequence(ks), self, self.close)

    def resize_maxrect(self, mult=0.9, aspect=1.0):
        """resize frame to a multiple of the maximum-size square

        mult: scale vs max
        aspect: width/height aspect ratio
        """
        self.move(0, 0)
        desktop = QDesktopWidget().availableGeometry()
        frame = self.frameGeometry()

        # set size to square that is mult of shorter side of desktop
        wmax = desktop.width() + self.width() - frame.width()
        hmax = desktop.height() + self.height() - frame.height()

        if aspect * hmax < wmax: # height-limited
            d1 = mult * hmax
            self.resize(d1 * aspect, d1)
        else: # width-limited
            d1 = mult * wmax
            self.resize(d1, d1 / aspect)

    def reset_plotview(self):
        """Reset the plot view (slightly different than autoscale)."""
        if not hasattr(self, "data"):
            return
        Ny, Nx = np.shape(self.data)
        self.plot1.setXRange(0, Nx, padding=0)
        self.plot1.setYRange(0, Ny, padding=0)

    def mouse_moved(self, evt):
        """update statusbar with x,y,z of mouse event"""
        if not self.plot1.sceneBoundingRect().contains(evt[0]):
            return
        if not hasattr(self, "data",):
            return

        mousePoint = self.plot1.vb.mapSceneToView(evt[0])
        x, y = int(mousePoint.x()), int(mousePoint.y())
        Ny, Nx = np.shape(self.data)
        if (0 < y < Ny) and (0 < x < Nx):
            z = "{:.4g}".format(self.data[y, x])
        else:
            z = "n/a"
        self.statusBar().showMessage(
            "x,y,z = {:.4g}, {:.4g}, {}".format(x, y, z))

    def openDialog(self):
        """Instantiate an Open File Dialog."""
        if hasattr(self, "opendialog_called"):
            startdir = None
        else:
            startdir = STARTDIR
            self.opendialog_called = True

        title = "{}: Open File".format(self.WINTITLE_PREFIX)
        filename = QFileDialog.getOpenFileName(self, title, startdir)[0]
        if filename:
            self.open(filename)

    def open(self, filename):
        """open image file and update displayed image data"""
        base, ext = osp.splitext(filename)
        if ext.lower() == '.hdf':
            hdat = read_hdf.Hdf(filename)
            A = np.rot90(hdat.sig, 2)
        elif ext[1:].lower() in ('tif', 'tiff', 'png', 'bmp', 'jpg', 'jpeg'):
            global sp_misc
            if sp_misc is None:
                import scipy.misc as sp_misc
            A = np.flipud(sp_misc.imread(filename, flatten=True))
        else:
            self.statusBar().showMessage("{} files currently unsupported".format(ext))
            return

        # TODO: enable toggling between linear and log imagescale
        if False:
            self.logdata = np.empty_like(A)
            self.logdata[A>0] = np.log10(A[A>0])
            self.logdata[A<=0] = np.nan
            self.data = self.logdata
        else:
            self.data = self.lindata = A

        self.img1.setImage(self.data)
        self.resize_maxrect(aspect=A.shape[1]/A.shape[0])
        self.reset_plotview()
        self.lut.move_to_parent()
        self.lut.setLevels()

        self.filename = filename
        self.statusBar().showMessage("opened file: {}".format(filename))
        self.statusbar_right.setText(osp.split(filename)[1])
        self.setWindowTitle("{}: {}".format(self.WINTITLE_PREFIX, filename))

    def helpDialog(self):
        if not hasattr(self, "help_window"):
            self.help_window = HelpWindow()
            self.help_window.move(self.x() + 24, self.y() + 32)
        self.help_window.show()

    def lutDialog(self):
        """Show the LUT Window, and arrange close to self."""
        self.lut.show()
        self.lut.move_to_parent()


class LUTWindow(QMainWindow):
    """Window for lookuptable"""
    def __init__(self, parent, *args, **kws):
        QMainWindow.__init__(self, parent=parent, *args, **kws)
        self.parent_window = parent
        self.setWindowTitle("LUT")
        self.setWindowIcon(QIcon(ICON_PATH))

        self.graphics_widget = pg.GraphicsLayoutWidget()
        self.setCentralWidget(self.graphics_widget)

        # Contrast/color control
        self.hist = pg.HistogramLUTItem()
        self.hist.setImageItem(self.parent_window.img1)
        self.graphics_widget.addItem(self.hist)
        self.hist.gradient.loadPreset('thermal') # pyqtgraph preset gradient

        # shortcuts
        QtGui.QShortcut(QtCore.Qt.Key_L, self, self.move_to_parent)

    def move_to_parent(self):
        parent_geom = self.parent_window.frameGeometry()
        self.move(parent_geom.x() + parent_geom.width(), parent_geom.y())
        self.resize(375, self.parent_window.height())

    def setLevels(self, data=None, nsigma=3):
        """Set color histogram levels to mean value +/- nsigma of data."""
        if data is None:
            data = self.hist.imageItem().image.view(np.ndarray)

        z0 = np.nanmean(data)
        dz = np.nanstd(data)
        zmin = max(np.nanmin(data), z0 - nsigma*dz)
        zmax = min(np.nanmax(data), z0 + nsigma*dz)
        self.hist.setLevels(zmin, zmax)


class HelpWindow(QTextEdit):
    """RyView Help Window"""
    def __init__(self, *args, **kws):
        QTextEdit.__init__(self, *args, **kws)
        # add various shortcuts to close the window
        QtGui.QShortcut(QtCore.Qt.Key_Escape, self, self.close)
        for ks in ("Ctrl+w", "Ctrl+F4", "Ctrl+q"):
            QtGui.QShortcut(QtGui.QKeySequence(ks), self, self.close)

        self.setWindowTitle("{} Help".format(OViewMainWindow.WINTITLE_PREFIX))
        self.setWindowIcon(QIcon(ICON_PATH))
        self.setText("{}\n\nversion: {}".format(__doc__, __version__))
        self.setReadOnly(True)
        self.resize(600, 400)


#==============================================================================
def main():
    app = QApplication(sys.argv)
    mywindow = OViewMainWindow()
    sys.exit(app.exec_())

if __name__ == "__main__":
    import oview
    main()


oview/io/read_hdf.py
"""
Read and parse hdf4 files produced at Omega-60 and Omega-EP.

author: J. R. Rygg
history:
    2017-05-19 basic support for p510_data summary hdf
    2017-04-13 created
"""

import os.path as osp
import sys

import numpy as np
import pyhdf.SD as sd

# lazy import: won't import until needed (gives tiny speedup on startup)
plt = None # import matplotlib.pyplot as plt


class Hdf():
    """Hdf file data from Omega-60 and Omega-EP data sources."""
    KNOWN_DATA_SOURCES = ('p510', 'asbo', 'sop', 'pxrdip') # known hdf file sources

    PXRDIP_SCAN_ORDER_DEFAULT = ('U', 'L', 'D', 'R', 'B') # default pxrdip scan order
    PXRDIP_DISPLAY_ORDER = { # display order for PXRDIP plates on omega-60 or ep
        'omega-60': ('R', 'D', 'L', 'U', 'B'),
        'omega-ep': ('L', 'U', 'R', 'D', 'B'),
    }

    def __init__(self, handle, pathlist=None, hint=None):
        self.handle = handle # filename or (TODO: implement) other keyword
        self.pathlist = pathlist # list of possible paths
        self.hint = hint # hint to preferred filename

        self.filename = self.handle#self.find_filename(self.handle, pathlist)
        self.title, self.fileext = osp.splitext(osp.split(self.filename)[1])

        self.facility = 'omega-ep' # TODO: omega-60 or NIF or other or None
        self.attributes = None
        self.datasets = None

        self.datasource = self.determine_datasource()
        self.read_file()

    def determine_datasource(self):
        """Determine data type of given file based on filename."""
        filename = self.filename.lower()
        for datasource in self.KNOWN_DATA_SOURCES:
            if datasource in filename:
                return datasource

    def read_file(self):
        """Read datafile at self.filename."""
        if self.fileext == '.hdf':
            filedata = sd.SD(self.filename, sd.SDC.READ)
            self.parse_filedata(filedata)
            print_hdf_info(filedata)
            filedata.end() # close hdf file

    def parse_filedata(self, filedata):
        self.attributes = filedata.attributes()
        self.datasets = filedata.datasets()

        if self.fileext == '.hdf' and self.datasource in self.KNOWN_DATA_SOURCES:
            parse_method = getattr(self, "parse_" + self.datasource)
            parse_method(filedata)

    def parse_p510(self, filedata):
        """Parse Omega p510 summary file (laser power results)."""
        self.shotnum = self.attributes['LOG_NUM']

        self.time = filedata.select('SUMMARY_DATA_TIME').get() # time [ns]
        self.power = filedata.select('SUMMARY_DATA_AVERAGE')
        self.beam_power = filedata.select('SUMMARY_DATA_UV').get()
#        self.beams = self.power.attributes()['BEAMS'] # list of beams
        self.power = self.power.get() # per-beam powers [TW]

    def parse_asbo(self, filedata):
        """Parse Omega ASBO hdf (Active Shock BreakOut, = Omega VISAR)."""
        dat = filedata.select("Streak_array")
        ref = filedata.select("Reference")
        att = self.attributes
        self.shotnum = ""
        self.location = att['Location']
        self.unit = att['Unit']

        att_d = dat.attributes()
        self.sweep_setting = att_d['SweepSpeed']
        self.timestamp = att_d['TimeStamp']

        A = dat.get().astype(float)
        self.sig = np.rot90(A[0] - A[1], 2)
        self.ref = np.rot90(ref.get().astype(float) - A[1], 2)
        self.bg = np.rot90(A[1], 2)

    def parse_sop(self, filedata):
        """Parse Omega SOP hdf (SOP = streaked optical pyrometer)."""
        dat = filedata.select("Streak_array")
        att = self.attributes
        self.shotnum = ""
        self.location = att['Location']
        self.unit = att['Unit']

        att_d = dat.attributes()
        self.sweep_setting = att_d['SweepSpeed']
        self.timestamp = att_d['TimeStamp']

        A = dat.get().astype(float)
        self.sig = np.rot90(A[0] - A[1], 2)
        self.bg = np.rot90(A[1], 2)

    def parse_pxrdip(self, filedata):
        """Parse PXRDIP hdf (Powder X-Ray Diffraction Image Plate)."""
        psl = filedata.select("PSL_per_px")
        psl_attr = psl.attributes()
        self.scan_img = filedata.select("PSL_per_px").get()
        self.shotnum = str(int(psl_attr['shotNum']))
        self.instrument = psl_attr['instrument'] # scanner
        self.scandelay = psl_attr['scanDelaySeconds']
        self.latitude = psl_attr['latitude']
        self.scale = psl_attr['pixelSizeX']
        self.scale_units = 'um per pixel'
        if psl_attr['pixelSizeY'] != self.scale:
            sys.stdout.write("anisotropic IP scan\n")
            print(self.scale, psl_attr['pixelSizeX'], psl_attr['pixelSizeY'])

        self.ip_scan_order = self.PXRDIP_SCAN_ORDER_DEFAULT
        self.ip_display_order = self.PXRDIP_DISPLAY_ORDER[self.facility]
        self.ijs = self.clip_pxrdip_plates()
        self.shuffle_pxrdip_plates(self.ijs)

    def clip_pxrdip_plates(self, di=40):
        """Return slices for different image plates.

        di = range in pixels over which to search for boundaries
        """
        scale = self.scale / 1000 # convert to mm
        L = 75 # mm approximate pxrdip ip length
        W = 50 # mm approximate pxrdip ip width

        yy = self.scan_img.mean(1) + 1e-3 # vertical lineout
        step = np.array((-1,-1,-1,1,1,1)) # step kernel
        j1 = np.argmax(np.convolve(yy, step, mode='same') / yy)
        jslice = slice(0, j1+5) # "vertical" slice

        xx = self.scan_img.mean(0) + 1e-3 # horizontal lineout
        i0 = int(L / scale) * np.array((1, 2, 3, 4, 5), dtype=int)
        i0[-1] -= int((L - W) / scale)
        xconv = np.convolve(xx, step, mode='same') / xx
        i1 = []
        for i in i0:
            imin = np.argmin(xconv[i-di:i+di])
            i1.append(i-di + np.argmax(xconv[i-di:i-di+imin]))

        ijs = [(jslice, slice(0, i1[0]))]
        ijs.extend([(jslice, slice(i1[k-1],i1[k])) for k in range(1,len(i1))])

        return ijs

    def shuffle_pxrdip_plates(self, ijs, pattern='all'):
        """Shuffle the pxrdip plates into desired order."""
        self.sig = B = np.zeros((2000, 1250))# + np.nan
        display_lookup = {ip: k for k, ip in enumerate(self.ip_display_order)}

        for k, ij in enumerate(ijs):
            Ak = self.scan_img[ij]
            Ny, Nx = np.shape(Ak)
            k2 = display_lookup[self.ip_scan_order[k]]
            if k < len(ijs) - 1:
                B[500*k2:500*k2+Ny,0:Nx] = Ak
            else:
                B[1500:1500+Ny,-Nx:] = Ak

    def imshow(self, **kws):
        """Generate matplotlib figure of image data."""
        global plt
        if plt is None:
            import matplotlib.pyplot as plt

        if self.datatype == 'pxrdip':
            figsize = (3.75, 6)
        else:
            figsize = (6, 6)

        A = self.sig
        kws['origin'] = 'lower'
        if 'cmap' not in kws:
            kws['cmap'] = 'viridis'
        if 'vmin' not in kws and 'vmax' not in kws:
            z0 = np.mean(A)
            dz = np.std(A)
            kws['vmin'] = max(A.min(), z0 - 3*dz)
            kws['vmax'] = min(A.max(), z0 + 3*dz)
            kws['vmin'], kws['vmax'] = thresh_vlim(A, thresh=0.05, Nbin=2000)

        fig = plt.figure(figsize=figsize, dpi=250)
        fig.canvas.window().move(0, 0)
        ax = fig.add_axes((0,0,1,1))
        im = ax.imshow(np.rot90(A,2), **kws)

        if kws['cmap'] == 'viridis':
            textcolor = 'w'
        else:
            textcolor = 'k'

        if self.datatype == 'pxrdip':
            fig.text(0.02, 0.985, "PXRDIP - shot {}".format(self.shotnum),
                     ha='left', va='top', color=textcolor)
            cax = fig.add_axes((0.2, 0.51, 0.03, 0.24))
            fig.colorbar(im, cax=cax)
            cax.set_ylabel("signal (PSL)", color=textcolor)
            for item in cax.get_yticklabels():
                item.set_color(textcolor)
                item.set_fontsize('small')
        elif self.datatype == 'asbo':
            if kws['cmap'] == 'viridis':
                color = 'w'
            else:
                color = 'k'
            fig.text(0.02, 0.985, "{}".format(self.unit.split()[0]),
                     ha='left', va='top', color=color)
            cax = fig.add_axes((0.4, 0.97, 0.57, 0.02))
            fig.colorbar(im, cax=cax, orientation='horizontal')

            for item in cax.get_xticklabels():
                item.set_color(textcolor)
                item.set_fontsize('small')


def thresh_vlim(data, thresh=0.01, Nbin=200):
    """Calculate colormap vlim based on a histogram threshold.
    Excludes fraction on top and bottom given by scalar or 2-tuple thresh.
    """
    thresh = thresh*np.ones(2) if np.size(thresh) < 2 else thresh
    H, bin_edges = np.histogram(data[~np.isnan(data)], Nbin)
    vmin, vmax = bin_edges[0], bin_edges[-1]
    csH = H.cumsum() / np.nansum(H) # fractional cumulative histogram sum
    ilo = np.where(csH < thresh[0])[0]
    ihi = np.where(csH > 1.0 - thresh[1])[0]
    if len(ilo) > 0: vmin = bin_edges[np.nanmax(ilo) + 1]
    if len(ihi) > 0: vmax = bin_edges[np.nanmin(ihi) + 1]
    return vmin, vmax


def print_hdf_info(h):
    sow = sys.stdout.write
    sow("Number of (datasets, attributes) = {}\n".format(h.info()))
    print(h.datasets())
    print(h.attributes())
    for d in h.datasets():
        sow("---- {} ----\n".format(d))
        d = h.select(d)
        sow(str(d.info()))
        sow('\n')
        att = d.attributes()
        for k in sorted(att.keys()):
            sow("{}\t{}\n".format(k, att[k]))

